### PR TITLE
Save forward RNG outputs when the memory-budget partitioner forces recompute

### LINKED ISSUE #190717: [BUG] dropout gradient mask mismatch when memory budget is low
### Bug Description

When `graphsafe_rng_functionalization = True` is used with `activation_memory_budget < 1.0` (or `min_cut_rematerialization_partition` / activation checkpointing), AOTAutograd produces **gradient mask mismatch** for `Dropout`.
When the activation memory budget is 0.0, the forward dropout mask is not saved to HBM and is forced to be recomputed during the backward pass. However, because the recomputed backward dropout operation does not receive the original forward RNG state, it generates a different random mask, resulting in a gradient mismatch between eager and compiled execution.

### Reproduction

Colab Notebook: [link](https://github.com/haifeng-jin/Colabs/blob/main/torch_tpu_gh_2015.ipynb)

### Versions

wn_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities
Hypervisor vendor:                       KVM
Virtualization type:                     full
L1d cache:                               32 KiB (1 instance)
L1i cache:                               32 KiB (1 instance)
L2 cache:                                1 MiB (1 instance)
L3 cache:                                38.5 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0,1
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Vulnerable
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Mitigation; PTE Inversion
Vulnerability Mds:                       Vulnerable; SMT Host state unknown
Vulnerability Meltdown:                  Vulnerable
Vulnerability Mmio stale data:           Vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Vulnerable
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Vulnerable
Vulnerability Spectre v1:                Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:                Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Not affected; BHI: Vulnerable
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Vulnerable
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] intel-cmplr-lib-ur==2025.3.3
[pip3] intel-openmp==2025.3.3
[pip3] mkl==2025.3.1
[pip3] numpy==2.0.2
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.19.0.56
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.28.9
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] nvtx==0.2.15
[pip3] onemkl-license==2025.3.1
[pip3] optree==0.19.1
[pip3] tbb==2022.3.1
[pip3] tcmlib==1.5.0
[pip3] torch==2.11.0+cu128
[pip3] torchao==0.10.0
[pip3] torchaudio==2.11.0+cu128
[pip3] torchcodec==0.11.0+cu128
[pip3] torchdata==0.11.0
[pip3] torchsummary==1.5.1
[pip3] torchtune==0.6.1
[pip3] torchvision==0.26.0+cu128
[pip3] triton==3.6.0
[pip3] umf==1.0.3
[conda] Could not collect

cc @ezyang @gchanan @kadeng @msaroufim @soulitzer @pbelevich @chauhang @penguinwu @bdhirsh @bobrenjc93 @aorenste @zou3519

### LINKED ISSUE #190758: Memory-budget partitioner recomputes RNG ops with fresh randomness in backward, silently corrupting gradients at any activation_memory_budget < 1.0
### 🐛 Describe the bug

When `torch.compile` is used with `torch._functorch.config.activation_memory_budget < 1.0`
and the model contains RNG ops (dropout, `randn_like`, ...), the partitioner can force
those ops to be recomputed in the backward graph, where they draw fresh randomness.
The backward then applies a different dropout mask than the forward actually used,
so gradients are silently wrong. No error, no warning.

This extends and corrects #190717, which reported the `budget == 0` case:

- `graphsafe_rng_functionalization` is not causally involved. It is only consulted
  inside `functionalize_rng_ops`, which never runs on this path: it is gated on
  user activation-checkpointing recompute tags (`partitioners.py:247-251, 4289-4293`),
  which partitioner-forced recompute never sets. The failure reproduces identically
  with the flag off, and on CPU.
- It is not only `budget == 0`. The `budget == 0` shortcut
  (`partitioners.py:3471-3472`) returns early and skips `solve_min_cut`'s
  "random op" recompute ban entirely — but at `0 < budget < 1` the knapsack path
  can *re-allow* recompute of the banned rng ops (`get_recomputable_banned_nodes`
  has no rng filter and `dont_ban` overrides the ban). I verified wrong gradients
  at budgets 0.0, 0.005 and 0.5.

### Minimal reproducer

Deterministic, and avoids conflating this bug with the documented eager-vs-inductor
Philox stream divergence — it checks the backward against the mask the compiled
forward itself applied:

```python
import torch
import torch.nn.functional as F
import torch._functorch.config as fc

fc.activation_memory_budget = 0.0  # also broken at 0.005 and 0.5

w = torch.arange(1.0, 26.0).view(5, 5)

def f(x):
    return F.dropout(x @ w, 0.5, True)

compiled = torch.compile(f)
torch.manual_seed(42)
x = torch.ones(2, 5, requires_grad=True)
out = compiled(x)
out.sum().backward()

mask = (out.detach() != 0).float()   # the mask the forward actually applied
expected = 2.0 * mask @ w.t()        # gradient implied by that mask
print(torch.equal(x.grad, expected)) # main: False — the backward redrew the mask
```

### Ablation

Triggers with the default inductor backend and with `backend="eager"`; with
`torch._inductor.config.fallback_random` True and False; at budgets 0.0, 0.005
and 0.5 (not at 1.0); on CPU builds and (per #190717's report) CUDA builds; also
with `torch.utils.checkpoint` (`use_reentrant=False`) combined with a low budget
(out-of-region rng ops), and with `randn_like` even in `eval()` mode.
`graphsafe_rng_functionalization` on/off makes no difference.

### Graph evidence

With `TORCH_LOGS=aot_graphs` at budget=0, the forward AND backward graphs each
draw their own independent seeds:

```
===== Forward graph 0 =====
    inductor_seeds_default: "i64[1][1]cpu" = torch.ops.prims.inductor_seeds.default(1, device(type='cpu'))
===== Backward graph 0 =====
    inductor_seeds_default: "i64[1][1]cpu" = torch.ops.prims.inductor_seeds.default(1, device(type='cpu'))
```

Full `TORCH_TRACE` log attached. With `fallback_random=True` the same happens
with `aten.native_dropout` recomputed directly in the backward.

### Training-level impact

Small two-dropout MLP, synthetic classification task, 5 seeds, CPU:

| configuration | mean eval acc | final train loss |
|---|---|---|
| eager | 0.894 | 0.224 |
| compiled, budget=0, with the fix below | 0.894 (identical curves with `fallback_random=True`) | 0.224 |
| compiled, budget=0, current main | 0.55–0.61 (near chance) | **NaN** |

### Fix

I have a two-part fix validated locally (CPU; the partitioner logic is
device-independent, CUDA validation via CI):

1. In the `budget == 0` shortcut, additionally save the tensor outputs of
   required-forward rng ops (op-level filter; all `getitem` projections for
   tuple-output ops). On the default path this saves only the small
   `inductor_seeds` tensor.
2. Exclude rng ops from the knapsack's tradeable candidates in
   `get_recomputable_banned_nodes`, so the solver's existing random-op ban stays
   final at every budget — as it already is at budget 1.0.

PR incoming with CPU regression tests (mask self-consistency at budgets 0 and
0.005, exact eager parity under `fallback_random=True`, and a backward-only-rng
case). Threading real RNG state through partitioner-forced recompute (extending
`functionalize_rng_ops` beyond AC tags) would be the fuller redesign — happy to
leave that call to maintainers.

The `budget == 0` shortcut predates graphsafe (#126318, May 2024).

[aot_graphs_buggy.log](https://github.com/user-attachments/files/30251903/aot_graphs_buggy.log)
[torch_trace_190717_buggy.log](https://github.com/user-attachments/files/30251902/torch_trace_190717_buggy.log)

### Error logs

No error is raised — that is the problem. Training runs normally and gradients
are silently wrong (backward applies a different dropout mask than the forward).
In our training test this collapsed accuracy to near chance with NaN loss.

### Versions

PyTorch version: 2.14.0a0+gitb761270
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A
OS: Ubuntu 22.04.5 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 4.4.0
Libc version: glibc-2.35
Python version: 3.10.12 (main, Jun 22 2026, 18:55:27) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-1057-oracle-aarch64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
cuDNN version: No CUDA

cc @soulitzer @pbelevich @chauhang @penguinwu @bdhirsh @bobrenjc93 @aorenste

### PR DESCRIPTION
**Pull request by HussainNizamani**
below is the more refined submission, please read carefully,
P.S: Fable-Assisted

Fixes #190758, Fixes #190717

## Problem

With `torch._functorch.config.activation_memory_budget < 1.0`, the partitioner can
force RNG ops (dropout, `randn_like`) to be recomputed in the backward graph. The
recomputed op draws fresh randomness there, so the backward applies a different
dropout mask than the one the forward actually used. Gradients are silently wrong —
no error, no warning. In a small training test at budget=0 this diverges to NaN loss
and near-chance accuracy; with the fix, compiled training matches eager exactly.
Full analysis, minimal repro, graph traces and training numbers: #190758.

## Root cause

Two independent paths produce the same corruption:

1. At `budget == 0`, `choose_saved_values_set` returns early with only the graph
   inputs and never calls `solve_min_cut`, so the existing "random op" recompute ban
   never runs. With `TORCH_LOGS=aot_graphs` the forward and backward graphs each
   contain their own `prims.inductor_seeds` call.
2. At `0 < budget < 1`, the ban does run, but the knapsack path can trade the banned
   rng ops back into recompute: `get_recomputable_banned_nodes` has no rng filter,
   and `dont_ban` overrides the ban. Reproduced at budgets 0.005 and 0.5.

The rng state-pairing machinery (`functionalize_rng_ops`) helps in neither case:
it is gated on user activation-checkpointing recompute tags, which partitioner-forced
recompute never sets.

## Fix

- The `budget == 0` shortcut now also saves the tensor outputs of forward rng ops.
  Ops are selected with the partitioner's existing `is_rng_op` (the
  `nondeterministic_seeded` tag): this matches `prims.inductor_seeds` and the aten
  rng ops, but not `inductor_lookup_seed`/`inductor_random`, which are deterministic
  given the seeds — so on the default inductor path only the small seeds tensor is
  saved. For tuple-output ops (`native_dropout`) all getitem projections are saved
  (saving the tuple node breaks saved-tensor accounting); backward-only rng ops are
  excluded (forcing them into saved values makes an invalid node a forward output).
- `get_recomputable_banned_nodes` no longer offers rng ops as knapsack candidates,
  so the solver's ban stays final at every budget, as it already is at budget 1.0.

Memory cost, stated honestly: on the default inductor path this saves one small
`inductor_seeds` tensor. On paths where aten rng ops survive to partition time
(`fallback_random=True`, `aot_eager`), it saves the rng ops' actual outputs — the
same tensors budget=1.0 already saves for them. Correctness over savings.

An alternative design — threading real RNG state through partitioner-forced
recompute by extending `functionalize_rng_ops` beyond AC tags — would preserve more
of the memory savings, but is a larger, CUDA-entangled change. This PR takes the
minimal-blast-radius repair; the state-threading redesign is left as a possible
follow-up for maintainers to weigh.

## Verified

- Structural tests capture the partitioned graphs and assert the backward contains
  no rng op while the forward saves the rng outputs (each dropout mask at both
  forcing budgets, every projection at budget 0), with a non-vacuousness guard
  proving the budget really forced recompute. Both fail without the fix.
- Behavioral tests: dropout fw/bw mask self-consistency at budgets 0.0 and 1.0,
  exact eager parity under `fallback_random=True` at budgets 0.0, 0.005 (knapsack
  regime) and 0.5, and a backward-only rng op asserting bit-exact gradients via a
  seed-aligned eager comparison. The consistency and parity tests fail without the
  fix.
- CUDA twins for the helper-based tests, so GPU CI exercises the same checks on
  real GPUs.
- Saved-value sets at budget=1.0 verified byte-identical before/after the change.
- Training check at budget=0: a small two-dropout MLP goes from NaN loss and
  near-chance accuracy to matching eager exactly (numbers in #190758).

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section

**Comment by HussainNizamani** [comment]
Thanks for the detailed review. I tested the coverage question on my machine — here is what I found and what I changed. P.S: Fable-Assisted

> **The coverage gap is real — confirmed, with one correction.** The review is right about the mechanism: the denylist solve only checked `op_types.is_random`, which is a hardcoded list of 3 ops (`native_dropout`, `rand_like`, `randn_like`). RNG ops that carry the `nondeterministic_seeded` tag but are not in that list were never banned, and this PR's knapsack filter never saw them. Two findings from actually running it:
> - `aten.bernoulli` cannot hit this path in practice: a core decomposition turns it into `torch.rand(...) < p` before the partitioner runs, and `rand` is covered.
> - `aten.multinomial` has no decomposition and does hit it — even in default config, not only with `aggressive_recomputation=True`. At `activation_memory_budget=0.5` the solver reaches the denylist tier, multinomial gets recomputed in the backward, and 2853 of 4096 gradient elements come out wrong (max abs diff 6.0) while the forward outputs stay identical. The same happens for `poisson` and `exponential` when they reach the partitioner.
>
> Since this is exactly the bug class this PR is about, I extended it: the denylist branch of `should_ban_recomputation` now also checks the tag-based `is_rng_op` (one line), so all budget paths use the same op classification. I also added multinomial regression tests: the backward graph must not contain multinomial, and gradients must match eager, in both config modes.
>
> One extra observation from stress-testing: the knapsack stage can un-ban compute-intensive ops, so for RNG ops the tag check (plus the existing knapsack filter) is the only hard guarantee. A cheap follow-up hardening would be to move the `is_rng_op` check above the allowlist/denylist split, so a tagged op that ever lands on the allowlist cannot slip through. Happy to do that here or in a separate PR.
>
> **On device-generic tests:** agreed — I moved this PR's numerics tests into a new device-generic class (`TestPartitioningRngBudget`) registered with `instantiate_device_type_tests` and removed all the manual `_cuda` twins.
>
> Extra verification on CPU beyond the unit tests: an op × budget × config matrix and a 9-budget sweep over a model with dropout + multinomial + randn_like show zero tagged RNG ops in any backward graph; compiled and eager gradients match bit-for-bit with `fallback_random`; a 60-step training run matches eager step by step, while a build with the new check reverted diverges from step 2; and partitioning decisions are byte-identical with the check disabled whenever there is no budget or no RNG op in the graph. Memory cost of the ban is exactly the saved RNG outputs (for multinomial, its int64 index tensor). CUDA-only paths (fused attention with dropout, cudnn RNN, `functionalize_rng_ops`) are covered by the graph-level argument here and by CI.
